tools/odbc/parameter_wrapper.cpp

```diff
@@ -0,0 +1,151 @@
+#include "parameter_wrapper.hpp"
+#include "duckdb_odbc.hpp"
+#include "duckdb/common/types/decimal.hpp"
+
+using duckdb::Decimal;
+using duckdb::idx_t;
+using duckdb::ParameterDescriptor;
+using duckdb::ParameterWrapper;
+
+ParameterWrapper::~ParameterWrapper() {
+}
+
+void ParameterWrapper::Clear() {
+	param_descriptors.clear();
+	cur_paramset_idx = 0;
+}
+
+void ParameterWrapper::ResetParamSetIndex() {
+	cur_paramset_idx = 0;
+}
+
+SQLRETURN ParameterWrapper::GetValues(std::vector<Value> &values) {
+	values.clear();
+	if (param_descriptors.empty()) {
+		return SQL_NO_DATA;
+	}
+	D_ASSERT(cur_paramset_idx < paramset_size);
+	// Fill values
+	for (idx_t desc_idx = 0; desc_idx < param_descriptors.size(); ++desc_idx) {
+		// set a proper parameter value
+		auto ret = param_descriptors[desc_idx].SetValue(cur_paramset_idx);
+		if (param_status_ptr) {
+			param_status_ptr[cur_paramset_idx] = ret;
+		}
+		if (ret != SQL_PARAM_SUCCESS) {
+			std::string error_msg;
+			error_msg.reserve(100);
+			auto len = snprintf((char *)error_msg.data(), 100,
+			                    "Error setting parameter value: ParameterSet '%ld', ParameterIndex '%ld'",
+			                    cur_paramset_idx, desc_idx);
+			error_msg.resize(len);
+			error_messages->emplace_back(error_msg);
+			if (!param_status_ptr) {
+				return SQL_ERROR;
+			}
+		}
+		values.emplace_back(param_descriptors[desc_idx].values[cur_paramset_idx]);
+	}
+	++cur_paramset_idx;
+	if (cur_paramset_idx == paramset_size) {
+		return SQL_NO_DATA;
+	}
+	return SQL_SUCCESS;
+}
+
+SQLRETURN ParameterDescriptor::ValidateNumeric(int precision, int scale) {
+	if (precision < 1 || precision > Decimal::MAX_WIDTH_DECIMAL || scale < 0 || scale > Decimal::MAX_WIDTH_DECIMAL ||
+	    scale > precision) {
+		// TODO we should use SQLGetDiagField to register the error message
+		std::string msg = "Numeric precision: " + std::to_string(precision) + " and scale: " + std::to_string(scale) +
```

Comment:
`StringUtil::Format` should simplify this construction, e.g.:

```c
auto msg = StringUtil::Format("Numeric precision %d and scale %d not supported", precision, scale);
```

Reply:
It is good to know. Using `StringUtil::Format` is  better, for sure.